README.md
# React Interval Update

[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)

## Installation

```
npm i --save react-interval-update
```

## Usage

### As a HOC

```jsx
import { withInterval } from 'react-interval-update';

const CurrentTimeWithUpdater = ({ onIntervalChange, interval }) => (
  <div>
    {new Date().toString()}
    <br />
    Update interval:
    <input
      value={interval}
      onChange={({ target }) => onIntervalChange(parseInt(target.value, 10))}
    />
  </div>
);

export default withInterval(CurrentTimeWithUpdater, { interval: 5000, prop: 'update' });
```

### As a component

```jsx
import { Interval } from 'react-interval-update';

const CurrentTime = () => new Date().toString();

return (
  <Interval interval={1000}>
    <CurrentTime />
  </Interval>
);
```


CHANGELOG
# Changelog

## Unreleased

## 0.2.0

* Setting interval to a negative integer cancels the timeout

## 0.1.0

* Package created


test/spec/Interval.js
/* globals jest */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-enzyme';
import { Interval } from '../../lib/index';

Enzyme.configure({ adapter: new Adapter() });

jest.useFakeTimers();

describe('Interval', () => {
  let component;
  let mockProps;
  let currentTime;

  const TimeComponent = () => (currentTime = new Date().getTime().toString());

  const getRequiredProps = () => ({
    interval: 1000
  });

  const setupComponent = (overrides = {}) => {
    mockProps = {
      ...getRequiredProps(),
      ...overrides
    };

    component = mount((
      <Interval {...mockProps}>
        <b><TimeComponent /></b>
        {null}
        Test
      </Interval>
    ));
  };

  beforeEach(setupComponent);

  it('renders normally', () => {
    expect(component.find('b')).toHaveText(currentTime);
  });

  it('sets timer matching interval', () => {
    expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1000);
  });

  describe('after timer completes', () => {
    let oldTime;

    beforeEach(() => {
      oldTime = currentTime;
      jest.runOnlyPendingTimers();
      component.update();
    });

    it('re-renders the child component', () => {
      expect(oldTime).not.toEqual(currentTime);
      expect(component.find('b')).toHaveText(currentTime);
    });

    describe('after timer completes again', () => {
      beforeEach(() => {
        oldTime = currentTime;
        jest.runOnlyPendingTimers();
        component.update();
      });

      it('re-renders the child component again', () => {
        expect(oldTime).not.toEqual(currentTime);
        expect(component.find('b')).toHaveText(currentTime);
      });
    });

    describe('when passed a new interval prop', () => {
      beforeEach(() => {
        component.setProps({ interval: 2000 });
        component.update();
      });

      it('calls setTimeout with new interval', () => {
        expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 2000);
      });
    });

    describe('when passed a negative interval prop', () => {
      beforeEach(() => {
        component.setProps({ interval: -1 });
        component.update();
      });

      it('does not call setTimeout', () => {
        expect(setTimeout).not.toHaveBeenCalledWith(expect.any(Function), -1);
      });
    });
  });
});


test/spec/withInterval.js
/* globals jest */
import React from 'react';
import Enzyme, { mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import 'jest-enzyme';
import { withInterval, Interval } from '../../lib/index';

Enzyme.configure({ adapter: new Adapter() });

jest.useFakeTimers();

describe('withInterval', () => {
  let component;

  class TestComponent extends React.Component {
    render() {
      return this.props.children;
    }
  }

  const WrappedTestComponent = withInterval(TestComponent, { interval: 1000, prop: 'update' });

  const setupComponent = (overrides = {}) => {
    component = mount((
      <WrappedTestComponent foo="bar">
        <b>Hello</b>
      </WrappedTestComponent>
    ));
  };

  beforeEach(setupComponent);

  it('renders normally', () => {
    expect(component.find('b')).toHaveText('Hello');
  });

  it('passes props to wrapped component', () => {
    expect(component.find(WrappedTestComponent)).toHaveProp('foo', 'bar');
  });

  it('renders Interval with the default interval', () => {
    expect(component.find(Interval)).toHaveProp('interval', 1000);
  });

  it('renders Component with the default interval', () => {
    expect(component.find(TestComponent)).toHaveProp('interval', 1000);
  });

  describe('when onIntervalChange called', () => {
    beforeEach(() => {
      component.find(TestComponent).prop('onIntervalChange')(2000);
      component.update();
    });

    it('renders Interval with the new interval', () => {
      expect(component.find(Interval)).toHaveProp('interval', 2000);
    });

    it('renders Component with the new interval', () => {
      expect(component.find(TestComponent)).toHaveProp('interval', 2000);
    });
  });
});


lib/Interval.js
// @flow
import * as React from 'react';

/** @memberOf Interval */
type Props = {
  /** Children to re-render */
  children: React.Node,
  /** Interval to update */
  interval: number,
  /** Prop to update to trigger a change */
  prop: string
};

export default class Interval extends React.Component<Props> {
  static defaultProps = {
    prop: 'key'
  };

  timer: TimeoutID;

  update = () => this.forceUpdate();

  render() {
    clearTimeout(this.timer);

    if (this.props.interval >= 0) {
      this.timer = setTimeout(this.update, this.props.interval);
    }

    return React.Children.map(
      this.props.children,
      (child, index) => React.isValidElement(child)
        ? React.cloneElement(child, {
          ...child.props,
          [this.props.prop]: [new Date().getTime(), index].join()
        }, child.props.children)
        : null
    );
  }
}


lib/withInterval.js
// @flow
import * as React from 'react';
import { Interval } from './index';

/** @memberOf IntervalWrapper */
type Props = {
  [string]: any
};

/** @memberOf IntervalWrapper */
type State = {
  interval: number
};

const withInterval = (
  Component: React.ComponentType<any>,
  intervalProps: { interval: number, [string]: any }
) =>
  class IntervalWrapper extends React.Component<Props, State> {
    state = {
      interval: intervalProps.interval
    };

    onIntervalChange = (interval: number): void => this.setState({ interval });

    render() {
      return (
        <Interval {...intervalProps} interval={this.state.interval}>
          <Component
            onIntervalChange={this.onIntervalChange}
            interval={this.state.interval}
            {...this.props}
          >
            {this.props.children}
          </Component>
        </Interval>
      );
    }
  };

export default withInterval;


stories/index.stories.js
import React from 'react';

import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { linkTo } from '@storybook/addon-links';

const specialAction = (name) =>
  (...args) => {
    console.log(name, args);
    return action(name)(...args);
  };

import { Interval, withInterval } from '../lib/index';

const CurrentTime = () => new Date().toString();

storiesOf('Interval', module)
  .addDecorator(story => (
    <div style={{ height: '100vh', width: '100vw' }}>
      {story()}
    </div>
  ))
  .add('update every second', () => (
    <Interval interval={1000}>
      <CurrentTime />
    </Interval>
  ))
  .add('update every 5 seconds', () => (
    <Interval interval={5000}>
      <CurrentTime />
    </Interval>
  ));

const CurrentTimeWithUpdater = ({ onIntervalChange, interval }) => (
  <div>
    {new Date().toString()}
    <br />
    Update interval: <input value={interval} onChange={({ target }) => onIntervalChange(parseInt(target.value, 10))} />
  </div>
);

const CurrentTimeWithInterval = withInterval(CurrentTimeWithUpdater, { interval: 5000, prop: 'update' });

storiesOf('withInterval', module)
  .addDecorator(story => (
    <div style={{ height: '100vh', width: '100vw' }}>
      {story()}
    </div>
  ))
  .add('default update every 5 seconds', () => (
    <CurrentTimeWithInterval />
  ));
